src/main/java/org/ethereum/beacon/discovery/database/ExpirationSet.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.ethereum.beacon.discovery.database;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+import kotlin.Pair;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Set with data expiring in configured time
+ *
+ * @param <V> data type
+ */
+public class ExpirationSet<V> implements Set<V> {
+
+  private final long expirationDelayMillis;
+  private final Clock clock;
+  private final long size;
+
+  TreeSet<Pair<Long, V>> dataExpiration =
+      new TreeSet<>(
+          (o1, o2) -> {
+            int expirationDeltaSignum = Long.signum(o1.getFirst() - o2.getFirst());
+            if (expirationDeltaSignum != 0) {
+              return expirationDeltaSignum;
+            }
+            return Integer.compare(o1.getSecond().hashCode(), o2.getSecond().hashCode());
```

Comment:
This comparison isn't safe as `hashCode` can return the same value for different objects. I suspect we should rework this to just store the node id which is a `Bytes` and implements `Comparable` itself.  Then the Compartor could be created with:
```
new TreeSet<>(
          Comparator.comparingLong(Pair<Long, Bytes>::getFirst)
              .reversed()
              .thenComparing(Pair::getSecond));
```

Though you should double check if the `.reversed()` should be there or not. :)

Reply:
Yeah, actually there could be disintegration  between `data` and `dataExpiration` in rare case and this will cause whole structure to be buggy